### PR TITLE
Set up the Association

### PR DESCRIPTION
**Pull request by Shaheryar0054**
- Created a class Classroom with the following:
  @label instance variable should be initialized in the constructor.
   Setter and getter for @label (remember about attr_accessor).
- Created the has-many/belongs-to relationship between Classroom and Student. The following should be implemented:
   Create the has-many side (a classroom has many students).
   Create the belongs-to side (a student belongs to a classroom).
   Make sure that when adding a student to a classroom it also sets the classroom for the student.
   Make sure that when setting the classroom for a student it also adds it to the classrooms' students.

- Created a class Book with the following:
  @title and @author instance variables should be initialized in the constructor.
  Setters and getters for instance variables (remember about attr_accessor).
- Created a class Rental with the following:
  @date instance variable should be initialized in the constructor.
  Setter and getter for @date (remember about attr_accessor).
- Created the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class 
  Rental. The following should be implemented:
  Create the has-many side of Book and Rental (a book has many rentals).
  Create the belongs-to side of Rental and Book (a rental belongs to a book).
  Create the has-many side of Person and Rental (a person has many rentals).
  Create the belongs-to side of Rental and Person (a rental belongs to a person).
  Modify the constructor of Rental so Book and Person are set in it.

### CODE REVIEW

File: classroom.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+  def initialize(label)
+    @label = label
+    @students = []
+  end
+
+  def add_student(student)
+    @students << student
+    student.classroom = self
+  end
```
Comment (nahnahdev) [inline classroom.rb]: - [ ] Great job so far @Shaheryar0054 There are just a few changes I will suggest you implement and you'll be on your way to the next Milestone :wink: 
- [ ] You did an incredible job creating the `'has-many' side (a classroom has many students)` in the `classroom.rb` file. But in the `student.rb` file, you also need to ensure that there is a `belongs-to relationship` between `Classroom` and `Student` :+1:  so kindly create a method that serves as the `'belongs-to' side (a student belongs to a classroom)` :wink: 
- [ ] That way, when setting the classroom for a student it also adds it to the classrooms' students :+1: 

- Example implementation 
```ruby
class Student < Person
  attr_reader :classroom
  
  def classroom=(classroom)
      @classroom = classroom
      classroom.students.push(self) unless classroom.students.include?(self)
  end
end
```

File: book.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals << rental
+  end
+end
```
Comment (nahnahdev) [inline book.rb]: - [ ] You have done an impressive job so far. However, per the project requirements, you needed to:

> Create the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. The following should be implemented:
> Create the has-many side of Book and Rental (a book has many rentals)

- Currently, this isn't being implemented properly in the `Book class`. I will suggest you ensure that the book itself is sent as a parameter to create a rental thereby achieving the `has-many` association. :wink: 
- Here is a [link](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship%20def%20add_rental(person,%20date)) to an example of this implementation in a previous lesson :+1: 
- Example implementation

```ruby
  def add_rental(person, date)
    Rental.new(date, self, person)
  end
```

File: test_person_rental_book.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# Test associations between person, rental and book. Sample code and output below:
+# plese note: This will be removed in the final version of the project as it defined to check ...
+# correct implementation of the associations bewteen person, rental and book.
+require './person'
```
Comment (nahnahdev) [inline test_person_rental_book.rb]: - [ ] Just as I mentioned in my previous comment, a similar implementation should be present in this `Person class` as well :+1: 
- [ ]  Per the project requirements, you needed to:

> Create the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. The following should be implemented:
> Create the has-many side of Person and Rental (a person has many rentals).

- Currently, this isn't being implemented in the `Person class`. I will suggest you include `a method for adding rentals `in the class. The person itself should be sent as a parameter to create the new rental thereby achieving the `has-many` association. :wink: 
- Here is a [link](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship%20def%20add_rental(person,%20date)) to an example of this implementation in a previous lesson :+1: 
```ruby
  def add_rental(book, date)
     Rental.new(date, book, self)
  end
```

**Review by nahnahdev — CHANGES_REQUESTED** [review]
## Project Status: CHANGES REQUIRED :recycle: 
Hi @Shaheryar0054   , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights :fire: 
- Professional README :heavy_check_mark: 
- All required files present :heavy_check_mark: 
- Associations almost completed :heavy_check_mark: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

You can also consider:
- N/A

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: person.rb [hunk 4]
```diff
@@ -9,9 +9,14 @@ def initialize(age, name = 'Unknown', parent_permission: true)
...
     @parent_permission = parent_permission
+    @rentals = []
   end
   attr_accessor :name, :age
-  attr_reader :id
+  attr_reader :id, :rentals
+
+  def add_rental(rental)
+    @rentals << rental
+  end
```
Comment (tugsem) [inline person.rb]: -  [ ] As the reviewer required, please pass the book, date, and person as parameters for this method. Then, you need to create a rental as the reviewer made for you. Add that rental to @rentals which belongs to a person. So, you need to give arguments (book, date, self) and create a new rental, pass that rental with `@rentals << rental`. 
- [ ] The same thing applies to books, in the add_rental method please add parameters(person, date, self) to create the rental and pass it to the rentals of the book.

File: rental.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  attr_accessor :date
+  attr_reader :book, :person
+
+  def initialize(book, person, date)
+    @book = book
+    @person = person
+    @date = date
+
+    book.add_rental(self)
+    person.add_rental(self)
```
Comment (tugsem) [inline rental.rb]: - [ ] Add_rental method should take different parameters for each class(person, book). So you should use
 `book.rentals << self` 
`person.rentals << self` instead this block.

File: rental.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  def book=(book)
+    @book = book
+    book.add_rental(self)
```
Comment (tugsem) [inline rental.rb]: - The same thing applies here. I suggest changing this line to `book.rentals << self`

File: rental.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  def person=(person)
+    @person = person
+    person.add_rental(self)
```
Comment (tugsem) [inline rental.rb]: - The same thing applies here. I suggest changing this line to `person.rentals << self`

File: person.rb (hunk 4 above)
Comment (tugsem) [inline person.rb]: - [ ] As the reviewer addressed before, please pass the person as a parameter to a new rental object. So you need to create a rental here. Pass (book, date, self) as an argument to create a rental. And add that rental to person's rentals with `@rentals << rental`

File: book.rb [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals << rental
+  end
```
Comment (tugsem) [inline book.rb]: - [ ] As the reviewer addressed before, please pass the book as a parameter to a new rental object. So you need to create a rental here. Pass (person, date, self) as an argument to create a rental. And add that rental to the book's rentals with `@rentals << rental`

**Review by tugsem — CHANGES_REQUESTED** [review]
## STATUS: REQUIRED CHANGES ♻️
Hello @Shaheryar0054 👋 , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are some issues that you still need to work on to go to the next project but you are almost there!

Alt Text

 ## Highlights 🚀
✔️Good job on adding 'rentals' object to the person class

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by mikethreels — APPROVED** [review]
Hi Shaheryar, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- good job following the instructions of the previous CR
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**